In [1]:
import pandas as pd
import geopandas as gpd
import fiona
import leafmap.foliumap as leafmap
import time
import requests
from geopy.geocoders import GoogleV3
from geopy.extra.rate_limiter import RateLimiter
import os

In [2]:
path_row = "data/fiber_lines.gpkg"
path_customers = "data/coverage_streets_raw.gpkg"

Load layers in `Right of Way` file

In [3]:
# Load all layers from right_of_way.gpkg
layers = fiona.listlayers(path_row)
gdf = [gpd.read_file(path_row, layer=lyr) for lyr in layers]
layers

['Area_1', 'Area_2', 'Area_3', 'Area_4']

Merge all the layers in the `Right of Way` package

In [4]:
# Merge them into one GeoDataFrame
gdf_right = gpd.GeoDataFrame(pd.concat(gdf, ignore_index=True), crs=gdf[0].crs)

Read `Customer coverage` data

In [5]:
# Load customer coverage lines
gdf_all = gpd.read_file(path_customers)

# Make sure they’re in the same CRS
gdf_all = gdf_all.to_crs(gdf_right.crs)

- Right of Way has 699 LineStrings
- Total Customer Coverage has 4522 LineStrings

In [6]:
gdf_all["geometry"]

0       MULTILINESTRING ((539051.588 734446.242, 53906...
1       MULTILINESTRING ((539348.629 735273.988, 53930...
2       MULTILINESTRING ((538783.354 735536.111, 53877...
3       MULTILINESTRING ((538652.132 734970.537, 53872...
4       MULTILINESTRING ((538667.543 735013.782, 53867...
                              ...                        
4517    MULTILINESTRING ((546468.623 711304.651, 54648...
4518    MULTILINESTRING ((547245.666 712127.567, 54724...
4519    MULTILINESTRING ((546565.5 711536.769, 546563....
4520    MULTILINESTRING ((546356.942 711486.078, 54635...
4521    MULTILINESTRING ((540988.433 711969.593, 54104...
Name: geometry, Length: 4522, dtype: geometry

Break down MultiLineStrings into LineStrings so that the lines that can iterated individually

In [7]:
gdf_right = gdf_right.explode(index_parts=False, ignore_index=True)
gdf_all = gdf_all.explode(index_parts=False, ignore_index=True)

Check for lines that intersect between the Right of Way and Customer Coverage and filter them out
- The lines may be really close and not perfectly intersect, so we need a buffer for 5 metres

In [8]:
# Option 1: exact geometry comparison

buffered_right = gdf_right.buffer(39)  # 5 meters tolerance
# gdf_non_right = gdf_all[~gdf_all.intersects(buffered_right.union_all())]
gdf_non_right = gdf_all[~gdf_all.intersects(gdf_right.unary_union)]
gdf_non_right["geometry"].value_counts().sum()

np.int64(2967)

There are 2967 LineStrings without Right of Way
- 4522 -> 2967() -> 2844(buffer of 5m) -> 2594(buffer off 39m)

Calculate distance and assign values to a field

In [9]:
gdf_non_right = gdf_non_right.to_crs(epsg=32631)
gdf_non_right["distance_m"] = gdf_non_right.geometry.length

In [14]:
gdf_non_right

id  code        fclass                                  name   ref  \
1        65  5122   residential                 Alhaji Moshoba Street  None   
2        89  5122   residential                                  None  None   
3       113  5122   residential                  Alhaji Bakare Street  None   
4       118  5122   residential                          Odobo Street  None   
5       120  5122   residential                                  None  None   
...     ...   ...           ...                                   ...   ...   
4516  67143  5121  unclassified  Prince Olanrewaju A. Elegushi Street  None   
4517  67152  5141       service                                  None  None   
4518  67158  5141       service                                  None  None   
4519  67176  5153       footway                                  None  None   
4522  67267  5122   residential                          Oniru Street  None   

        rid  count  customer_count  \
1        65      1               1   
2        89      2               2   
3       113      1               1   
4       118      2               2   
5       120      1               1   
...     ...    ...             ...   
4516  67143      1               1   
4517  67152      1               1   
4518  67158      3               3   
4519  67176      1               1   
4522  67267      3               3   

                                               geometry  distance_m  
1     LINESTRING (539348.629 735273.988, 539302.595 ...  957.132449  
2     LINESTRING (538783.354 735536.111, 538772.826 ...  190.338778  
3     LINESTRING (538652.132 734970.537, 538728.93 7...  206.279081  
4     LINESTRING (538667.543 735013.782, 538670.207 ...  197.334658  
5     LINESTRING (538492.971 734787.325, 538514.186 ...  190.186096  
...                                                 ...         ...  
4516  LINESTRING (555030.164 711400.444, 555307.009 ...  277.661891  
4517  LINESTRING (546196.109 711369.032, 546198.249 ...   76.756600  
4518  LINESTRING (546468.623 711304.651, 546481.238 ...   69.261008  
4519  LINESTRING (547245.666 712127.567, 547241.028 ...    9.006087  
4522  LINESTRING (540988.433 711969.593, 541047.955 ...  152.937416  

[2967 rows x 10 columns]

Plot the Customer Coverage without Right of Way

In [ ]:
# Create a map centered roughly on your data
m = leafmap.Map(center=[6.45, 3.39], zoom=9, style="streets")

# Style for non-right-of-way lines
style_non_right = {
    "color": "blue",
    "weight": 2,
}

# Style for right-of-way lines
style_right = {
    "color": "red",
    "weight": 2,
}

# Add the GeoDataFrame
# tooltip_fields = [col for col in ["road_street_name", "LOCAL GOVERNMENT", "distance(m)"] if col in gdf_non_right.columns]

# Add non-right-of-way roads
m.add_gdf(
    gdf_non_right,
    layer_type="line",
    layer_name="Non-Right-of-Way Roads",
    style=style_non_right,
)

# Add right-of-way roads
m.add_gdf(
    gdf_right,
    layer_type="line",
    layer_name="Right-of-Way Roads",
    style=style_right,
)

# Zoom to fit both datasets
m.zoom_to_gdf(pd.concat([gdf_non_right, gdf_right]))

m

Browser

In [ ]:
# webbrowser.open("map.html")

Save File

In [ ]:
#

Reverse geocode lines to get coordinates

In [18]:
API_KEY = os.getenv("banjo_google_api_key")
print(API_KEY)

AIzaSyBjm5CbJQiZFeW3IcrOpT5SnGqavRtSJhM


In [19]:
geolocator = GoogleV3(api_key=API_KEY, timeout=10)
# Set to False to raise exceptions on failures (useful for debugging)
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=5, swallow_exceptions=False)
reverse_geocode = RateLimiter(geolocator.reverse, min_delay_seconds=5, swallow_exceptions=False )

In [20]:
def reverse_geocode(lat, lon, api_key):
    """
    Reverse geocode coordinates using Google Maps API.
    Returns the formatted address, or None if not found.
    """
    try:
        url = f"https://maps.googleapis.com/maps/api/geocode/json?latlng={lat},{lon}&key={api_key}"
        response = requests.get(url)
        result = response.json()

        if result["status"] == "OK" and len(result["results"]) > 0:
            address = result["results"][0]["formatted_address"]
            print(f"Address found: {address}")
            return address
        else:
            print(f"No address found for {lat}, {lon}")
            return None
    except Exception as e:
        print(f"Error reverse geocoding ({lat}, {lon}): {e}")
        return None

In [21]:
def batch_reverse_geocode(gdf, api_key, batch_size=50, delay=2):
    """
    For each LineString in the GeoDataFrame, compute centroid,
    reverse geocode it with Google API, and add results as new columns.
    """
    # Ensure centroids are computed
    gdf["centroid"] = gdf.geometry.centroid
    gdf["lat"] = gdf.centroid.y
    gdf["lon"] = gdf.centroid.x

    addresses = []

    for i, row in gdf.iterrows():
        lat, lon = row["lat"], row["lon"]
        address = reverse_geocode(lat, lon, api_key)  # <- your function
        addresses.append(address)

        # Pause after a batch to avoid quota issues
        if (i + 1) % batch_size == 0:
            print(f"Processed {i+1} rows, pausing {delay}s...")
            time.sleep(delay)

    gdf["address"] = addresses
    return gdf

In [ ]:
if gdf_non_right.crs != "EPSG:4326":
    gdf_non_right = gdf_non_right.to_crs(epsg=4326)

gdf_non_right_with_addr = batch_reverse_geocode(gdf=gdf_non_right, api_key=API_KEY)

Address found: Unity Estate, 6 Aderinola Olamiju Cl, Ojodu, Lagos 101233, Lagos, Nigeria
Address found: 8 Joseph Odunlami Street, Ojodu, Lagos 101233, Lagos, Nigeria
Address found: ALHAJI BAKARE STR OTA, Ifako-Ijaiye, Lagos 101233, Lagos, Nigeria
Address found: 6 Adenle, Odobo St, Ojodu, Lagos 101233, Lagos, Nigeria
Address found: 9 Otuyelu St, Ojodu, Lagos 100218, Lagos, Nigeria
Address found: 5a Akintoye Cl, Ojodu, Lagos 101233, Lagos, Nigeria
Address found: 16 Osunkeye St, Ojodu, Lagos 101232, Lagos, Nigeria
Address found: Harmony Estate, 29 Kolapo Boluwade Cres, Ifako-Ijaiye, Lagos 101232, Lagos, Nigeria
Address found: 5 Cajeoleforo St, Ojodu, Lagos 101232, Lagos, Nigeria
Address found: 18 Col. Asieluwe St, Ojodu, Lagos 101232, Lagos, Nigeria


In [ ]:
from difference_intersecting_lines import filter_non_intersecting_lines

gdf = filter_non_intersecting_lines("data/fiber_lines.gpkg", "data/coverage_streets_raw.gpkg")